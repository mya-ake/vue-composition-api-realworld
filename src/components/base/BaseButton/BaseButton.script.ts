import { createComponent, reactive, computed } from '@vue/composition-api';
import { BaseButtonType } from './BaseButton.type';

export const useButton = (clickFunc: BaseButtonType.Props['onClick']) => {
  const localState = reactive({ processing: false });
  const disabled = computed(() => localState.processing);

  const onClick = async (event: Event) => {
    localState.processing = true;
    await clickFunc(event);
    localState.processing = false;
  };

  return {
    disabled,
    onClick,
  };
};

export default createComponent({
  props: {
    type: {
      default: 'button',
    },
    onClick: {
      default: () => {},
    },
  },
  setup(props: BaseButtonType.Props) {
    const { disabled, onClick } = useButton(props.onClick);
    return {
      disabled,
      clickListener: onClick,
    };
  },
});
