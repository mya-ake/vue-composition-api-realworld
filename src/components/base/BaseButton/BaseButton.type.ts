export namespace BaseButtonType {
  export interface Props {
    type: 'button' | 'submit';
    onClick: (event: Event) => void | Promise<void>;
  }
}
