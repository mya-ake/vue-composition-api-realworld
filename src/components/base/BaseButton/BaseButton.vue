<template>
  <button class="btn" :type="type" :disabled="disabled" @click="clickListener">
    <slot />
  </button>
</template>

<script lang="ts" src="./BaseButton.script.ts"></script>
