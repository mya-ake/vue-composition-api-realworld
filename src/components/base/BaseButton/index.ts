export { default as BaseButton } from './BaseButton.vue';
export * from './BaseButton.type';
