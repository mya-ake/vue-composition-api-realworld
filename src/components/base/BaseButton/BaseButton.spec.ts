import { shallowMount, Wrapper } from '@vue/test-utils';
import BaseButton from './BaseButton.vue';
import { useButton } from './BaseButton.script';

describe('components/base/BaseButton', () => {
  describe('mountable', () => {
    let wrapper: Wrapper<Vue>;

    beforeEach(() => {
      wrapper = shallowMount(BaseButton, {});
    });

    it('is vue instance', () => {
      expect(wrapper.isVueInstance()).toBe(true);
    });

    it('snapshot', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('useButton', () => {
    describe('onClick', () => {
      it('onClickの関数が実行中はdisabledになる', () => {
        const onClickMock = jest
          .fn()
          .mockReturnValue(new Promise(resolve => resolve()));
        const { disabled, onClick } = useButton(onClickMock);
        onClick({} as any);
        expect(disabled.value).toBe(true);
      });
    });
  });
});
