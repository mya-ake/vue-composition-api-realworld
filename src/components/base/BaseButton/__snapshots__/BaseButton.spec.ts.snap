// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/base/BaseButton mountable snapshot 1`] = `
<button
  class="btn"
  type="button"
/>
`;
